### PR TITLE
Add a code to generate camera definition file

### PR DESCRIPTION
**Pull request by SaliniVenate**
A code to generate camera definition file by querying the camera device is added to test folder. Code has dependency on tinyxml which can be installed using:
$ sudo apt install libtinyxml-dev

To run the code:
$ cd camera-streaming-daemon/tools/export_v4l2_param_xml
$ make
$./export_v4l2_param -d <device name with path> -f <output filename>
   eg: ./export_v4l2_param -d /dev/video0 -f  camera_def.xml

### CODE REVIEW

**Comment by SaliniVenate** [comment]
@anselmolsm  Makefile is listed in .gitignore but I added it to the commit using `git add -f`. What is the correct way of doing it- should I create a Makefile.am instead of Makefile in or add another target to existing Makefile.am?

**Comment by anselmolsm** [comment]
@SaliniVenate you could either deal with Makefile.am or change .gitignore from `Makefile` to `/Makefile`, i.e. only ignore the file generated at the root dir of the project.

**Comment by SaliniVenate** [comment]
@anselmolsm @hamishwillee Can you please review the code?

File: tools/export_v4l2_param_xml/export_v4l2_param_xml.cpp [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+struct v4l2_capability cap;
+struct v4l2_query_ext_ctrl query_ext_ctrl;
+struct v4l2_queryctrl queryctrl;
+
+struct options {
+    const char *filename;
+    const char *devicename;
+};
+
+std::string str_toupper(std::string s);
```
Comment (hamishwillee) [inline tools/export_v4l2_param_xml/export_v4l2_param_xml.cpp]: Normally items should be sensibly named. Not sure what "troupper" means

File: tools/export_v4l2_param_xml/export_v4l2_param_xml.cpp [hunk 2]
```diff
@@ -0,0 +1,451 @@
+#include <algorithm>
```
Comment (hamishwillee) [inline tools/export_v4l2_param_xml/export_v4l2_param_xml.cpp]: Nice to have some explanation header and copyright info?

**Comment by hamishwillee** [comment]
I've added a few comments, but my C++ reviewing skills really aren't good enough for proper review. What testing have you done of the output?

**Comment by SaliniVenate** [comment]
@hamishwillee, These are the steps I followed for testing:
1) Hosted the xml file generated by the tool using a http server.
2) Updated conf file of csd and started csd.
3) Opened QGC and checked whether QGC is able to parse the xml file  and display the parameters.

**Comment by hamishwillee** [comment]
@SaliniVenate Sounds like a good basic test - particularly since the file isn't validated by CSD for XML correctness or anything else AFAIK. 

Have you checked that the parameters extracted are the full set that that can/should be extracted? Essentially I want to be sure that file generated:
- Contains entries all parameters that can be obtained from a v4l2 camera (ie not hard coded to some subset).
- That the output is XML compliant and compliant with the standard
- Perhaps basic checking that the values aren't obvious garbage for any parameters?

@anselmolsm You should do the technical review.

File: tools/export_v4l2_param_xml/export_v4l2_param_xml.cpp [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+}
+
+static int parse_argv(int argc, char *argv[], struct options *opt)
+{
+    static const struct option options[] = {{"device-name with path", required_argument, NULL, 'd'},
+                                            {"output file name", required_argument, NULL, 'f'}};
+    int c;
+
+    if (argc != 5) {
+        printf("Usage: ./export_v4l2_param_xml -d  <devide name along with path>  - f <output "
```
Comment (hamishwillee) [inline tools/export_v4l2_param_xml/export_v4l2_param_xml.cpp]: device name?
Should this have a file extension? (the example doesn't)

**Comment by SaliniVenate** [comment]
@hamishwillee 
- The tool currently queries the camera device to get a list of parameters that can be set by the user. The list of parameters in xml file are not a hardcoded subset of v4l2 parameters, but a subset whose setting is supported by the device.
- I've added a check in the code to make sure that generated xml file is a valid xml file. If not it should print appropriate error message.
- The values for parameters are queried from camera device and so it is safe to assume that it is not garbage.

**Comment by lbegani** [comment]
@SaliniVenate Is there anything pending to close this?

**Comment by SaliniVenate** [comment]
@lbegani I believe no. I've  taken care of all the review comments and made changes accordingly.

**Pull request merged**